File: src/leetCodeProblems/Stacks/StackRawImpl.go
package main

//import "log"

type Stack []string

func (s *Stack) IsEmpty() bool {
	return len(*s) == 0
}
func (s *Stack) Push(str string) {
	*s = append(*s, str)
}
func (s *Stack) Len() int {
	return len(*s)
}
func (s *Stack) Pop() (string, bool) {
	if s.IsEmpty() {
		return "", false
	} else {
		index := len(*s) - 1
		elem := (*s)[index]
		*s = (*s)[:index]
		return elem, true
	}
}

// func main() {

// 	stack := new(Stack)
// 	stack.Push("One")
// 	stack.Push("Two")
// 	log.Println(stack.Pop())
// 	log.Println(stack.Pop())
// }

File: src/leetCodeProblems/Stacks/StackRawImpl.java
package leetCodeProblems.StacksAndQueues;

/**
 * Reference - https://www.educba.com/stack-vs-queue/
 * @author anshul.agrawal
 *
 */
public class StackRawImpl {
	private int maxSize;
	private String[] stackArray;
	private int top;

	public StackRawImpl(int s) {
		maxSize = s;
		stackArray = new String[maxSize];
		top = -1;
	}

	public void push(String j) {
		top += 1;
		stackArray[top] = j;
	}

	public String pop() {
		String temp = stackArray[top];
		top -= 1;
		return temp;
	}

	public String peek() {
		return stackArray[top];
	}

	public boolean isEmpty() {
		return (top == -1);
	}

	public boolean isFull() {
		return (top == maxSize - 1);
	}

}
